boa_engine/src/object/shape/shared_shape.rs

```diff
@@ -0,0 +1,381 @@
+use indexmap::{IndexMap, IndexSet};
+use rustc_hash::FxHashMap;
+
+use crate::property::{Attribute, PropertyDescriptor, PropertyKey};
+
+use super::{PropertyDescriptorAttribute, Shape, Slot, TransitionKey, TransitionType};
+
+/// TODO: doc
+#[derive(Debug, Trace, Finalize)]
+struct Inner {
+    /// Cached property map, only constucted when a property is accesed through it.
+    #[unsafe_ignore_trace]
+    property_table: RefCell<FxHashMap<PropertyKey, (u32, PropertyDescriptorAttribute)>>,
+
+    forward_property_transitions: GcRefCell<FxHashMap<TransitionKey, WeakGc<Inner>>>,
+    previous: Option<SharedShape>,
+
+    #[unsafe_ignore_trace]
+    property_key: PropertyKey,
+    property_index: u32,
+
+    #[unsafe_ignore_trace]
+    property_attributes: PropertyDescriptorAttribute,
+
+    // TODO: add prototype to shape
+    transition_type: TransitionType,
+}
+
+#[derive(Debug, Trace, Finalize, Clone)]
+pub(crate) struct SharedShape {
+    inner: Gc<Inner>,
+}
+
+impl SharedShape {
+    const DEBUG: bool = true;
+
+    #[inline]
+    pub(crate) fn previous(&self) -> Option<&SharedShape> {
+        self.inner.previous.as_ref()
+    }
+    #[inline]
+    pub(crate) fn property_key(&self) -> &PropertyKey {
+        &self.inner.property_key
+    }
+    #[inline]
+    pub(crate) fn property_index(&self) -> u32 {
+        self.inner.property_index
+    }
+    fn transition_type(&self) -> TransitionType {
+        self.inner.transition_type
+    }
+    #[inline]
+    pub(crate) fn is_root(&self) -> bool {
+        self.inner.previous.is_none()
+    }
+    #[inline]
+    fn get_forward_transition(&self, key: &TransitionKey) -> Option<WeakGc<Inner>> {
+        let map = self.inner.forward_property_transitions.borrow();
+        map.get(key).cloned()
+    }
+
+    fn new(inner: Inner) -> Self {
+        Self {
+            inner: Gc::new(inner),
+        }
+    }
+
+    #[inline]
+    pub(crate) fn root() -> Self {
+        Self::new(Inner {
+            property_table: RefCell::default(),
+            forward_property_transitions: GcRefCell::new(FxHashMap::default()),
+            property_key: PropertyKey::Index(0),
+            property_index: 0,
+            property_attributes: PropertyDescriptorAttribute::default(),
+            previous: None,
+            transition_type: TransitionType::Insert,
+        })
+    }
+
+    fn create_property_transition(&self, key: TransitionKey) -> Self {
+        if Self::DEBUG {
+            println!(
+                "Shape: Create Transition: {} -------> {}",
+                self.property_key(),
+                key.property_key
+            );
+        }
+        let next_property_index = if self.previous().is_none() {
+            self.property_index()
+        } else {
+            self.property_index() + 2
+        };
+        let new_inner_shape = Inner {
+            property_table: RefCell::default(),
+            forward_property_transitions: GcRefCell::new(FxHashMap::default()),
```

Comment:
Not sure what you mean? If you are talking about`GcRefCell` and `RefCell` they don't do any computation.

BTW the latest benchmarks aren't correct (they look horrible) because I have a `DEBUG` bool flag in `SharedShape` that I forgot to turn off before pushing that prints a lot of  shape state to stdout.